lib/blocktree/blocktree.go

```diff
@@ -297,29 +298,51 @@ func (bt *BlockTree) Leaves() []Hash {
 	return la
 }
 
-// HighestCommonAncestor returns the highest block that is a Ancestor to both a and b
-func (bt *BlockTree) HighestCommonAncestor(a, b Hash) (Hash, error) {
-	bt.RLock()
-	defer bt.RUnlock()
-
-	an := bt.getNode(a)
-	if an == nil {
+// LowestCommonAncestor returns the lowest common ancestor between two blocks in the tree.
+func (bt *BlockTree) LowestCommonAncestor(a, b Hash) (Hash, error) {
+	// Get nodes to iterate through
+	aNode := bt.getNode(a)
+	if aNode == nil {
 		return common.Hash{}, ErrNodeNotFound
 	}
 
-	bn := bt.getNode(b)
-	if bn == nil {
+	bNode := bt.getNode(b)
+	if bNode == nil {
 		return common.Hash{}, ErrNodeNotFound
 	}
-
-	ancestor := an.highestCommonAncestor(bn)
-	if ancestor == nil {
-		// this case shouldn't happen - any two nodes in the blocktree must
-		// have a common ancestor, the lowest of which is the root node
-		return common.Hash{}, fmt.Errorf("%w: %s and %s", ErrNoCommonAncestor, a, b)
+	return lowestCommonAncestor(aNode, bNode)
+}
+func lowestCommonAncestor(aNode, bNode *node) (Hash, error) {
+	// balance out block height to compare
+	aNum := aNode.number
+	bNum := bNode.number
+	if aNum > bNum {
+		diff := aNum - bNum
+		if diff > aNum {
+			return common.Hash{}, errors.New("out of bounds ancestor check")
+		}
+		for diff > 0 {
+			aNode = aNode.parent
+			diff--
+		}
+	} else if bNum > aNum {
+		diff := bNum - aNum
+		if diff > bNum {
+			return common.Hash{}, errors.New("out of bounds ancestor check")
+		}
+		for diff > 0 {
+			bNode = bNode.parent
+			diff--
+		}
 	}
 
-	return ancestor.hash, nil
+	for {
+		if aNode.hash == bNode.hash {
+			return aNode.hash, nil
+		}
+		aNode = aNode.parent
+		bNode = bNode.parent
+	}
```

Comment:
Add a check when the parent is nil so we don't go above the root, that would panic

Reply:
Im not sure this is a possibility. Can add it in anyways, but in what case would this be possible? It would have to be us processing a block that isn't a part of the chain I think ...

Reply:
the old code used a for loop and had the check just as the break conditions. I added this check, but do not think it's needed or will every be hit. This is also why there is no test case.

In typing this, deciding I will add a panic, that we can remove later if we want, just in case I am wrong and we somehow we hit this case. However, I do not think we can or will